jobs/cloud_replicate.Jenkinsfile

```diff
@@ -0,0 +1,147 @@
+             defaultValue: "",
+             trim: true)
+    ] + pipeutils.add_hotfix_parameters_if_supported()),
+    durabilityHint('PERFORMANCE_OPTIMIZED')
+])
+
+def build_description = "[${params.STREAM}]"
+
+// Reload pipecfg if a hotfix build was provided. The reason we do this here
+// instead of loading the right one upfront is so that we don't modify the
+// parameter definitions above and their default values.
+if (params.PIPECFG_HOTFIX_REPO || params.PIPECFG_HOTFIX_REF) {
+    node {
+        pipecfg = pipeutils.load_pipecfg(params.PIPECFG_HOTFIX_REPO, params.PIPECFG_HOTFIX_REF)
+        build_description = "[${params.STREAM}-${pipecfg.hotfix.name}]"
+    }
+}
+
+// no way to make a parameter required directly so manually check
+// https://issues.jenkins-ci.org/browse/JENKINS-3509
+if (params.VERSION == "") {
+    throw new Exception("Missing VERSION parameter!")
+}
+
+// runtime parameter always wins
+def cosa_img = params.COREOS_ASSEMBLER_IMAGE
+cosa_img = cosa_img ?: pipeutils.get_cosa_img(pipecfg, params.STREAM)
+def basearches = params.ADDITIONAL_ARCHES.split() as List
+basearches = basearches ?: pipecfg.additional_arches
+
+// we always release for x86_64
+basearches += 'x86_64'
+// make sure there are no duplicates
+basearches = basearches.unique()
+
+def stream_info = pipecfg.streams[params.STREAM]
+
+build_description += "[${basearches.join(' ')}][${params.VERSION}]"
+currentBuild.description = "${build_description} Waiting"
+
+// We just lock here out of an abundance of caution in case somehow
+// two jobs run for the same stream, but that really shouldn't happen.
+// Also lock version-arch-specific locks to make sure these builds are finished.
+lock(resource: "cloud-replicate-${params.STREAM}") {
+    cosaPod(cpu: "1", memory: "512Mi", image: cosa_img) {
+    try {
+
+        currentBuild.description = "${build_description} Running"
+
+        def s3_stream_dir = pipeutils.get_s3_streams_dir(pipecfg, params.STREAM)
+        def gcp_image = ""
+        def ostree_prod_refs = [:]
+
+        // Fetch metadata files for the build we are interested in
+        stage('Fetch Metadata') {
+            def ref = pipeutils.get_source_config_ref_for_stream(pipecfg, params.STREAM)
+            pipeutils.shwrapWithAWSBuildUploadCredentials("""
+            cosa init --branch ${ref} ${pipecfg.source_config.url}
+            cosa buildfetch --build=${params.VERSION} \
+                --arch=all --url=s3://${s3_stream_dir}/builds \
+                --aws-config-file \${AWS_BUILD_UPLOAD_CONFIG}
+            """)
+        }
+
+        def builtarches = shwrapCapture("""
+                          cosa shell -- cat builds/builds.json | \
+                              jq -r '.builds | map(select(.id == \"${params.VERSION}\"))[].arches[]'
+                          """).split() as Set
+        assert builtarches.contains("x86_64"): "The x86_64 architecture was not in builtarches."
```

Comment:
Minor: feels like we could split this out into a shared function.

Reply:
yeah maybe later.